src/psyclone/psyir/nodes/assignment.py

```diff
@@ -188,16 +191,31 @@ def reference_accesses(self, var_accesses):
     @property
     def is_array_range(self):
         '''
-        returns: True if the lhs of the assignment is an array with at \
-            least one of its dimensions being a range and False \
-            otherwise.
+        returns: True if the lhs of the assignment is an array access with at \
+            least one of its dimensions being a range and False otherwise.
         rtype: bool
 
         '''
-        from psyclone.psyir.nodes import ArrayReference, Range
-        if not isinstance(self.lhs, ArrayReference):
-            return False
-        return any(dim for dim in self.lhs.children if isinstance(dim, Range))
+        # It's not sufficient simply to check for a Range node as that may be
+        # part of an argument to an Operator or function that performs a
+        # reduction and thus returns a scalar result, e.g. a(SUM(b(:))) = 1.0
+        # TODO #658 this check for reductions needs extending to also support
+        # user-implemented functions.
+        if isinstance(self.lhs, (ArrayReference, StructureReference)):
+            ranges = self.lhs.walk(Range)
+            for array_range in ranges:
+                opn = array_range.ancestor(Operation)
+                while opn:
+                    if opn.operator in REDUCTION_OPERATORS:
+                        # The current array range is in an argument to a
```

Comment:
I think this is an over-zealous check as a reduction operation may still return an array. For example, for the sum operation, if the second argument is a dimension value and the array has more than one dimension. If this is not an issue for nemo then I'm happy to leave it as it is as we can argue that it is an issue for #658, but I thought it best to raise it. This highlights two things to me. 1) the problem of just accepting Fortran intrinsics with all their options in the PSyIR; I think it would be better for us to distinguish between the different types of `sum` with different names in the PSyIR to make life simpler and 2) the issue of what to do if we don't know this information at code generation time (as I don't think we can always guarantee that we do).

Reply:
Ooh, that's difficult. I've see-sawed back and forth and come down in favour of delaying this to #658. I've added an x-failing test to illustrate the problem. I've also checked with the NEMO code base and this situation never occurs (for the particular configuration I'm building).

Reply:
I'm happy with that, thanks for adding the test.